src/irmin/node.ml

```diff
@@ -22,53 +22,81 @@ let src = Logs.Src.create "irmin.node" ~doc:"Irmin trees/nodes"
-          if equal_metadata m M.default then contents else contents_m m)
-    |~ case0 "node" `Node
-    |~ case0 "contents" (`Contents M.default)
-    |~ case1 "contents" M.t (fun m -> `Contents m)
-    |> sealv
+  type 'key contents_entry = { name : Path.step; contents : 'key }
+  [@@deriving irmin]
 
-  type entry = { kind : kind; name : P.step; node : H.t } [@@deriving irmin]
+  type 'key contents_m_entry = {
+    metadata : Metadata.t;
+    name : Path.step;
+    contents : 'key;
+  }
+  [@@deriving irmin]
 
-  let equal_entry_opt = Type.(unstage (equal [%typ: entry option]))
+  module StepMap = Map.Make (struct
+    type t = Path.step [@@deriving irmin ~compare]
+  end)
 
-  let to_entry (k, v) =
-    match v with
-    | `Node h -> { name = k; kind = `Node; node = h }
-    | `Contents (h, m) -> { name = k; kind = `Contents m; node = h }
+  type 'h node_entry = { name : Path.step; node : 'h } [@@deriving irmin]
 
-  let of_entry n =
-    ( n.name,
-      match n.kind with
-      | `Node -> `Node n.node
-      | `Contents m -> `Contents (n.node, m) )
+  type entry =
+    | Node of node_key node_entry
+    | Contents of contents_key contents_entry
+    | Contents_m of contents_key contents_m_entry
+    (* NOTE: the [_hash] cases are only externally reachable via
+       [Portable.of_node]. *)
+    | Node_hash of Hash.t node_entry
+    | Contents_hash of Hash.t contents_entry
+    | Contents_m_hash of Hash.t contents_m_entry
+  [@@deriving irmin]
 
-  module StepMap = Map.Make (struct
-    type t = P.step
+  type t = entry StepMap.t
+  type value = [ `Contents of contents_key * metadata | `Node of node_key ]
 
-    let compare = Type.(unstage (compare P.step_t))
-  end)
+  (* FIXME:  special-case the default metadata in the default signature? *)
+  let value_t =
+    let open Type in
+    variant "value" (fun n c x -> function
+      | `Node h -> n h
+      | `Contents (h, m) ->
+          if equal_metadata m Metadata.default then c h else x (h, m))
+    |~ case1 "node" node_key_t (fun k -> `Node k)
+    |~ case1 "contents" contents_key_t (fun h ->
+           `Contents (h, Metadata.default))
+    |~ case1 "contents-x" (pair contents_key_t Metadata.t) (fun (h, m) ->
+           `Contents (h, m))
+    |> sealv
 
-  type value = [ `Contents of hash * metadata | `Node of hash ]
-  type t = entry StepMap.t
+  let to_entry (k, (v : value)) =
+    match v with
+    | `Node h -> Node { name = k; node = h }
+    | `Contents (h, m) ->
+        if equal_metadata m Metadata.default then
+          Contents { name = k; contents = h }
+        else Contents_m { metadata = m; name = k; contents = h }
+
+  let of_entry : entry -> step * value = function
+    | Node n -> (n.name, `Node n.node)
+    | Contents c -> (c.name, `Contents (c.contents, Metadata.default))
+    | Contents_m c -> (c.name, `Contents (c.contents, c.metadata))
+    | Node_hash _ | Contents_hash _ | Contents_m_hash _ ->
+        (* Not reachable after [Portable.of_node] *)
```

Comment:
is there an invariant that `of_entry` is not called on the `[_hash]`?

Reply:
Yes. I've extended this comment to clarify somewhat.